Library/Homebrew/utils/io_selector.rb

```diff
@@ -0,0 +1,62 @@
+require "delegate"
+require "English"
+
+require "extend/io"
+
+module Utils
+  class IOSelector < DelegateClass(Hash)
+    attr_reader :separator
+
+    alias all_streams keys
+    alias all_tags values
+    alias tag_of fetch
+
+    def self.each_line_from(streams = {}, separator = $RS, &block)
+      new(streams, separator).each_line_nonblock(&block)
+    end
+
+    def initialize(streams = {}, separator = $RS)
+      super(streams.invert.compare_by_identity)
+      @separator = separator
+    end
+
+    def each_line_nonblock
+      each_readable_stream_until_eof do |stream|
+        line = stream.readline_nonblock(separator) || ""
+        yield(tag_of(stream), line)
+      end
+      close_streams
+    end
+
+    def pending_streams
+      @pending_streams ||= all_streams.dup
+    end
+
+    private
+
+    def each_readable_stream_until_eof(&block)
```

Comment:
Maybe `read_io_until_eofs` or `multiplex_streams_until_eof` or similar would be a better name? To me, in Ruby, a name `each_X` suggests iterating over the Xs, processing each once in turn. This method doesn't iterate over the streams; it multiplexes over them in an arbitrary order determined by external state, with possible repeats. (The `readable_streams.each` does iterate over some streams, but the streams in each call are an arbitrary sublist of the streams passed in to the `IOSelector` object, and doesn't correspond to the behavior visible to the caller.)